src/libstd/sys_common/gnu/libbacktrace.rs

```diff
@@ -121,12 +123,27 @@ pub fn print(w: &mut Write, idx: isize, addr: *mut libc::c_void,
     // case. There is no evidence at the moment to suggest that a more carefully
     // constructed file can't cause arbitrary code execution. As a result of all
     // of this, we don't hint libbacktrace with the path to the current process.
-    unsafe fn init_state() -> *mut backtrace_state {
+    //
+    // This can also return a File which should remain open until
+    // the first use of the state.
+    unsafe fn init_state() -> (*mut backtrace_state, Option<fs::File>) {
         static mut STATE: *mut backtrace_state = ptr::null_mut();
-        if !STATE.is_null() { return STATE }
-        STATE = backtrace_create_state(ptr::null(), 0, error_cb,
+        if !STATE.is_null() { return (STATE, None) }
+
+        let (filename, file) = match ::sys::backtrace::get_executable_filename() {
+            Ok((filename, file)) => {
+                // filename is purposely leaked here since libbacktrace requires
+                // it to stay allocated permanently
+                let filename_ptr = filename.as_ptr();
+                mem::forget(filename);
+                (filename_ptr, Some(file))
+            },
+            Err(_) => (ptr::null(), None),
+        };
+
+        STATE = backtrace_create_state(filename, 0, error_cb,
                                        ptr::null_mut());
-        STATE
+        (STATE, file)
```

Comment:
Ah ok, thanks for the explanation! I think it's probably easiest/best to just leak it here rather than special casing the result later on (depending on internal impl details of libbacktrace). We can probably use `.into_raw_handle()` here as well instead of `mem::forget`